iroh-util/src/lock.rs

```diff
@@ -0,0 +1,110 @@
+use anyhow::Result;
+use file_guard::{FileGuard, Lock};
+use std::fs::File;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+/// Manages a lock file used to track if an iroh program
+/// is already running.
+/// The lock is released either when the object is dropped
+/// or when the program stops.
+pub struct ProgramLock {
+    path: PathBuf,
+    lock: Option<FileGuard<Rc<File>>>,
+}
+
+impl ProgramLock {
+    /// Create a new lock for the given program. This does not yet acquire the lock.
+    pub fn new(prog_name: &str) -> Result<Self> {
+        let path = crate::iroh_data_path(&format!("{}.lock", prog_name))?;
+        Ok(Self { path, lock: None })
+    }
+
+    /// Check if the current program is locked or not.
+    pub fn is_locked(&self) -> bool {
+        if !self.path.exists() {
+            return false;
+        }
+
+        // Even if we manage to lock the file this won't last since the drop implementation
+        // of FileGuard releases the underlying lock.
+        if let Ok(file) = File::create(&self.path) {
+            file_guard::try_lock(&file, Lock::Exclusive, 0, 1).is_err()
+        } else {
+            false
+        }
+    }
+
+    /// Try to acquire a lock for this program.
+    pub fn acquire(&mut self) -> Result<()> {
+        let file = Rc::new(File::create(&self.path)?);
+
+        file_guard::lock(file, Lock::Exclusive, 0, 1)
+            .map(|lock| self.lock = Some(lock))
+            .map_err(|err| err.into())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn create_test_lock(name: &str) -> ProgramLock {
+        ProgramLock {
+            path: PathBuf::new().join(name),
+            lock: None,
+        }
+    }
+
+    #[test]
```

Comment:
This should be probably guarded by `cfg(unix)`, same for the `nix` dependency

Reply:
done in latest commit.